### PR TITLE
make : deprecate

### PR DESCRIPTION
**Pull request by ggerganov**
ref #10268 

Disabling the Makefile unless `LLAMA_MAKEFILE` is defined. This should give more focus for the CMake build and will help clean-up any usage of `make` in the existing CI. The Makefile will be eventually removed or stripped to a very minimal version.

### CODE REVIEW

**Comment by MaggotHATE** [comment]
> stripped to a very minimal version

That would be a much more preferred option. Like LostRuins mentioned in the original discussion, Makefile is a very simple and handy option that gives additional flexibility to projects that use llama.cpp, and it is multiplatform.

If you see the current state of Makefile as a burden, it is worth at least leaving a simplified version - for example, `main` and `server` as the most popular applications. This would serve not only as a quick "build and test" option, but also as an example for those who use Makefile in their projects. Then, for all advanced options users can be directed towards CMake only.

**Comment by slaren** [comment]
Is there a single platform that does not have easy access to cmake? If typing two commands instead of one is too much effort, we can always have the `Makefile` run `cmake`. In any case, let's keep that discussion to #10268.

**Comment by ggerganov** [comment]
Need to fix the `scripts/compare*` to use CMake builds instead of the Makefile and we can merge. The Swift build can be fixed later in #10525.

**Comment by max-krasnyansky** [comment]
Very nice! 
Good to see mostly deletions in the changes :)

**Review by slaren — APPROVED** [review]

File: Makefile [hunk 1]
```diff
@@ -1,3 +1,7 @@
+ifndef LLAMA_MAKEFILE
+$(error The Makefile build is deprecated. Use the CMake build instead)
```
Comment (slaren) [inline Makefile:2]: Might be good to link to https://github.com/ggerganov/llama.cpp/blob/master/docs/build.md here.
```suggestion
$(error The Makefile build is deprecated. Use the CMake build instead. For more details, see https://github.com/ggerganov/llama.cpp/blob/master/docs/build.md)
```

**Comment by slaren** [comment]
If you want a static build just disable `BUILD_SHARED_LIBS`. Do not use `make`, it will be removed completely in the future.

**Comment by jboero** [comment]
https://github.com/ggerganov/llama.cpp/pull/10663

**Pull request merged**